### PR TITLE
[Merged by Bors] - chore (RingTheory.Congruence): make RingCon extend AddCon and Con

### PR DESCRIPTION
**Pull request by mattrobball**
Currently `RingCon` extends `Add` and `Mul`. This changes it to extend `AddCon` and `Con` directly.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mattrobball** [comment]
!bench

**Comment by eric-wieser** [comment]
That doesn't look like good news to me

**Comment by mattrobball** [comment]
Total instructions over the whole library are essentially flat. 

The red one is small. 

But you are correct in that it doesn’t represent a performance improvement (yet). 

I think the main argument is stylistic consistency.

**Comment by mattrobball** [comment]
I am skeptical of most improvements beyond instructions broadly but this does seem better.

File: Mathlib/RingTheory/Congruence.lean [hunk 1]
```diff
@@ -76,27 +71,17 @@ section Basic
...
       rcases x with ⟨⟨x, _⟩, _⟩
       rcases y with ⟨⟨y, _⟩, _⟩
-      have : x = y := h
-      subst x; rfl }
+      congr!
+      rw [Setoid.ext_iff,(show x.Rel = y.Rel from h)]
+      simp}
 
-@[simp]
 theorem rel_eq_coe : c.r = c :=
```
Comment (eric-wieser) [inline Mathlib/RingTheory/Congruence.lean:85]: Why did this lose its `simp` attribute?

File: Mathlib/RingTheory/Congruence.lean [hunk 2]
```diff
@@ -34,17 +34,13 @@ Most of the time you likely want to use the `Ideal.Quotient` API that is built o
...
 which preserves both. -/
-structure RingCon (R : Type*) [Add R] [Mul R] extends Setoid R where
-  /-- Ring congruence relations are closed under addition -/
-  add' : ∀ {w x y z}, r w x → r y z → r (w + y) (x + z)
-  /-- Ring congruence relations are closed under multiplication -/
-  mul' : ∀ {w x y z}, r w x → r y z → r (w * y) (x * z)
+structure RingCon (R : Type*) [Add R] [Mul R] extends Con R, AddCon R where
 #align ring_con RingCon
 
+attribute [inherit_doc RingCon] RingCon.toCon RingCon.toAddCon
```
Comment (eric-wieser) [inline Mathlib/RingTheory/Congruence.lean:42]: Won't this give completely the wrong documentation?

Comment (mattrobball) [reply Mathlib/RingTheory/Congruence.lean:42]: I got confused as to why I needed to document this in the first place and just did that. What is a better solution?

File: Mathlib/RingTheory/Congruence.lean [hunk 3]
```diff
@@ -121,11 +106,6 @@ protected theorem mul {w x y z} : c w x → c y z → c (w * y) (x * z) :=
   c.mul'
 #align ring_con.mul RingCon.mul
 
-@[simp]
-theorem rel_mk {s : Setoid R} {ha hm a b} : RingCon.mk s ha hm a b ↔ Setoid.r a b :=
-  Iff.rfl
-#align ring_con.rel_mk RingCon.rel_mk
-
```
[the remark this answers is not in the document]
Comment (mattrobball) [reply Mathlib/RingTheory/Congruence.lean:128]: Ok. I will restore.

File: Mathlib/RingTheory/Congruence.lean (hunk 2 above)
Comment (eric-wieser) [reply Mathlib/RingTheory/Congruence.lean:42]: I think there's an `add_decl_doc` command or similar, with which you can attach the docstrings which were already in the file.

File: Mathlib/RingTheory/Congruence.lean (hunk 1 above)
Comment (mattrobball) [reply Mathlib/RingTheory/Congruence.lean:85]: LHS simplified (asymmetrically). Removed it and things seemed fine. Let me investigate further with the switching of order.

File: Mathlib/RingTheory/Congruence.lean (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattrobball) [reply Mathlib/RingTheory/Congruence.lean:128]: Do we want a `mk’ also?

[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/RingTheory/Congruence.lean:128]: We have this problem all over mathlib, so I don't think we need to create a different `mk` here. 

I'm tempted to revive the `FlatHack` branch for morphisms/subobjects/congruence relations only, which would remove this wart from the public API.

File: Mathlib/RingTheory/Congruence.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/RingTheory/Congruence.lean:85]: I think that's probably expected (and one of the things I dislike about new-style structures), though perhaps worth a comment

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**